lib/tapioca/generic_type_registry.rb

```diff
@@ -0,0 +1,178 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Tapioca
+  # This class responsible for storing and looking up information related to generic types.
+  #
+  # The class stores 2 different kinds of data, in two separate lookup tables:
+  #   1. a lookup of generic type instances by name: `@generic_instances`
+  #   2. a lookup of type variable serializer by constant and type variable
+  #      instance: `@type_variables`
+  #
+  # By storing the above data, we can cheaply query each constant against this registry
+  # to see if it declares any generic type variables. This becomes a simple lookup in the
+  # `@type_variables` hash table with the given constant.
+  #
+  # If there are no entries, then we can cheaply know that we can skip generic type
+  # information generation for this type.
+  #
+  # On the other hand, if we get a result, then the result will be a hash of type
+  # variable to type variable serializers. This allows us to associate type variables
+  # to the constant names that represent them, easily.
+  module GenericTypeRegistry
+    class TypeVariableSerializer
+      extend T::Sig
+      sig do
+        params(
+          type_variable_type: Symbol,
+          variance: Symbol,
+          fixed: T.untyped,
+          lower: T.untyped,
+          upper: T.untyped
+        ).void
+      end
+      def initialize(type_variable_type, variance, fixed, lower, upper)
+        parts = []
+        parts << ":#{variance}" unless variance == :invariant
+        parts << "fixed: #{fixed}" if fixed
+        parts << "lower: #{lower}" unless lower == T.untyped
+        parts << "upper: #{upper}" unless upper == BasicObject
+
+        parameters = parts.join(", ")
+
+        @str = T.let(type_variable_type.to_s, String)
+        @str += "(#{parameters})" unless parameters.empty?
+      end
+
+      sig { returns(String) }
+      def to_s
+        @str
+      end
+    end
+
+    @generic_instances = T.let(
+      {},
+      T::Hash[String, Module]
+    )
+
+    @type_variables = T.let(
+      {},
+      T::Hash[Integer, T::Hash[Integer, TypeVariableSerializer]]
+    )
+
+    class << self
+      extend T::Sig
+
+      # This method is responsible building the name of the instantiated concrete type
+      # and cloning the given constannt so that we can return a type that is the same
+      # as the current type but is a different instance and has a different name method.
+      #
+      # We cache those cloned instances by their name in `@generic_instances`, so that
+      # we don't keep instantiating a new type every single time it is referenced.
+      # For example, `[Foo[Integer], Foo[Integer], Foo[Integer], Foo[String]]` will only
+      # result in 2 clones (1 for `Foo[Integer]` and another for `Foo[String]`) and
```

Comment:
Shouldn't it also create a type for the tuple: `[Foo[Integer], Foo[Integer], Foo[Integer], Foo[String]]`?

Reply:
No, since the tuple is not generic.